test/server/guarddog_impl_test.cc

```diff
@@ -0,0 +1,156 @@
+  GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  auto pet_dog = gd.createWatchDog(1);
+  for (int i = 0; i < 6; i++) {
+    time_point_ += std::chrono::milliseconds(100);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    pet_dog->touch();
+    gd.forceCheckForTest();
+  }
+}
+
+TEST_F(GuardDogMissTest, MissTest) {
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(stats_store, config_miss, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  time_point_ += std::chrono::milliseconds(501);
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  gd.forceCheckForTest();
+  gd.stopWatching(unpet_dog);
+  unpet_dog = nullptr;
+}
+
+TEST_F(GuardDogMissTest, MegaMissTest) {
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(stats_store, config_mega, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  time_point_ += std::chrono::milliseconds(501);
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  gd.forceCheckForTest();
+  gd.stopWatching(unpet_dog);
+  unpet_dog = nullptr;
+}
+
+TEST(GuardDogBasicTest, StartStopTest) {
+  NiceMock<Stats::MockStore> stats;
+  NiceMock<Configuration::MockMain> config(0, 0, 0, 0);
+  NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+}
+
+TEST(GuardDogBasicTest, LoopIntervalNoKillTest) {
+  NiceMock<Stats::MockStore> stats;
+  NiceMock<Configuration::MockMain> config(40, 50, 0, 0);
+  NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+  EXPECT_EQ(gd.loopIntervalForTest(), 40);
+}
+
+TEST(GuardDogBasicTest, LoopIntervalTest) {
+  NiceMock<Stats::MockStore> stats;
+  NiceMock<Configuration::MockMain> config(100, 90, 1000, 500);
+  NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+  EXPECT_EQ(gd.loopIntervalForTest(), 90);
+}
+
+TEST(WatchDogBasicTest, ThreadIdTest) {
+  NiceMock<Stats::MockStore> stats;
+  NiceMock<Configuration::MockMain> config(100, 90, 1000, 500);
+  NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+  auto watched_dog = gd.createWatchDog(123);
+  EXPECT_EQ(watched_dog->threadId(), 123);
+  gd.stopWatching(watched_dog);
+}
+// If this test fails it is because the SystemTime object has become nontrivial
+// or we are compiling under a compiler and library combo that makes the
+// SystemTime object require a lock to be atomicly modified.
+//
+// The WatchDog/GuardDog relies on this being a lock free atomic for perf
+// reasons so some workaround will be required if this test starts failing.
+TEST(WatchDogTimeTest, AtomicIsAtomicTest) {
+  ASSERT(std::atomic<ProdSystemTimeSource>{}.is_lock_free());
```

Comment:
Is it the `ProdSystemTimeSource` or `SystemTime` you want to be atomic lock free here?